controller/src/main/java/io/pravega/controller/store/stream/records/StreamTruncationRecord.java

```diff
@@ -81,17 +82,44 @@
     private final boolean updating;
 
     @Builder
-    public StreamTruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
-                                  Set<Long> deletedSegments, Set<Long> toDelete, long sizeTill, boolean updating) {
-        this.streamCut = ImmutableMap.copyOf(streamCut);
-        this.span = ImmutableMap.copyOf(span);
-        this.deletedSegments = ImmutableSet.copyOf(deletedSegments);
-        this.toDelete = ImmutableSet.copyOf(toDelete);
+    /**
+     * This is a private constructor that is only directly used by the builder during the deserialization. 
+     * The deserialization passes @param copyCollections as false so that we do not make an immutable copy of the collection
+     * for the collection passed to the constructor via deserialization. 
+     *
+     * The all other constructors, the value of copyCollections flag is true and we make an immutable collection copy of 
+     * the supplied collection. 
+     * All getters of this class that return a collection always wrap them under Collections.unmodifiableCollection so that
+     * no one can change the data object from outside.  
+     */
+    private StreamTruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
+                                  Set<Long> deletedSegments, Set<Long> toDelete, long sizeTill, boolean updating, boolean copyCollections) {
+        this.streamCut = copyCollections ? ImmutableMap.copyOf(streamCut) : streamCut;
```

Comment:
All of this just makes the code more complicated. The constructor should have a contract and the callers should abide by it, making it conditional is just weird. If you really want to have it two ways you can overload the constructor, but honestly it seems like in such a case it's probably better to just make the copy.

Reply:
the constructor contract on public constructors is a simple one. 
This is a private constructor where we are doing the optimization. 

This whole thing is being done for reducing creation of avoidable collection.Node objects.